DOCS: Adding Collective Operations Semantics

manjugv opened the pull request:
The collective operation semantics are not documented. This PR is an effort to capture the current semantics briefly.

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
-\b Invocation semantics: The ucc\_collective\_init routine is a non-blocking collective operation to initialize the buffers, operation type, reduction type, and other information required for the collective operation. All participants of the team should call the initialize operation. The routine returns once the participants enter the collective initialize operation. The collective operation is i ...
+**Invocation semantics**: The ucc\_collective\_init routine is a non-blocking
+collective operation to initialize the buffers, operation type, reduction type,
+and other information required for the collective operation. All participants of
```
_The routine returns once the participants enter the collective initialize operation._

This is a strong statement - i never realized we have such a constraint. Does this mean that ucc_collective_init is potentially blocking, ie it says that the call "may not return" until everybody enter collective. I think this is incorrect.

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
 
-\b Collective type: The collective operation supported by UCC is defined by the enumeration ucc\_coll\_type\_t. It supports three types of collective operations: (a) UCC\_{ALLTOALL,ALLTOALLV, ALLGATHER, ALLGATHERV, ALLREDUCE, REDUCE_SCATTER, REDUCE_SCATTERV, BARRIER} operations where all participants contribute to the results and receive the results (b) UCC\_{REDUCE, GATHER, GATHERV, FANIN} where ...
+**Collective Type**: The collective operation supported by UCC is defined by the
+enumeration ucc\_coll\_type\_t. The semantics are briefly described here,
```
are -> is (briefly described)

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+enumeration ucc\_coll\_type\_t. The semantics are briefly described here,
+however in most cases it agrees with the semantics of collective operations in
+the popular programming models such as MPI and OpenSHMEM. When they differ, the
+semantics changes are documented. All collective operations operate on the team.
```
operate -> are executed (otherwise "operations operate")

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+however in most cases it agrees with the semantics of collective operations in
+the popular programming models such as MPI and OpenSHMEM. When they differ, the
+semantics changes are documented. All collective operations operate on the team.
+For the collective operations defined in the enumeration, all participants of
```
"participants participate"

Lets rephrase: All team members participate in collective operation.

First part "For the collective operations defined in the enumeration" is unnecessary imho, since we don't support anything else (which is not defined in enumeration)

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+semantics changes are documented. All collective operations operate on the team.
+For the collective operations defined in the enumeration, all participants of
+the team participate in the collective operations. Further the team should be
+created with endpoints, where the “eps” should be ordered and contiguous.
```
_Further the team should be created with endpoints, where the “eps” should be ordered and contiguous._

Is this a requirement? I thought that potentially we can support creation of teams with non-contig eps. What is meant here?

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+is designated as root.
 
-\b Ordering: The team can be configured for ordered collective operations or unordered collective operations. For unordered collectives, the user is required to provide the “tag”, which is an unsigned 64-bit integer. 
++ The UCC\_COLL\_TYPE\_BCAST operation moves the data from the src buffer of
```
lets remove "src" and "dest" words here, because it is a bit misleading. In coll_args for bcast we init args.src.info.buffer for ALL ranks, root and non root (args.dst.info is not initialized).

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+participants into the operation, it exits the operation completing it locally.
 
-In the non-synchronized collective model, on entry, the participants can read or write to other participants. If the input and output buffers are defined on the team and RMA operations are used for data transfer, it is the responsibility of the user to ensure the readiness of the buffer. On exit, the participants may exit once the read and write to the local buffers are completed. 
++ The UCC\_COLL\_TYPE\_FANIN gathers the data from all participants in the team
```
FANIN/FANOUT don't move data. These are essentially synchronizing primitives (like UCC_COLL_TYPE_BaRRIER). Fanin can be viewed as 0 byte reduce/gather, and fanout is 0 byte bcast/scatter.

manjugv replied:
I see. Small change to what you are saying - we don't have to write it as a 0-byte msg. Probably, we could write it as sync msg and leave it to the implementation, if it wants to use 0-byte or more that is up to the implementation.

I will rewrite the semantics. Let's see if it will capture the use cases.

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+to block_size * num\_participants. Here, the “count” represents the number of
+data elements. The
+"dt_elem_size" represents the size of the data element in bytes. The
+"num_participants" define the number of participants in the team. The data on
```
define -> defines

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+"dt_elem_size" represents the size of the data element in bytes. The
+"num_participants" define the number of participants in the team. The data on
+the root is placed in the receive buffer ordered by the “ep” ordering. For
+example, if the participants’ endpoints are ordered as “ep\_a” to “ep\_n”, the
```
How do we define "ordering" of non-contiguous EPs ? Say, ep is just a pointer - each process might have different ordering then (local pointers may differ).

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
++ In the UCC\_COLL\_TYPE\_SCATTER operation, the root participant of the
+collective operation sends data to all other participants. It sends the same
+amount of data (block_size) to all participants. The size of the block
+(block_size) is “dt_elem_size * count”. The total amount of data received by the
```
received -> sent

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+elements. The "dt_elem_size" represents the size of the data element in bytes.
+
++ The UCC\_COLL\_TYPE\_REDUCE\_SCATTER first performs a element-wise
+reduction on the src buffer and then scatters the result to the dst buffer. If
```
The sentence starts with "IF" but there is no "then"

manjugv replied:
Huh .. The If is an error there.

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+reduction on the src buffer and then scatters the result to the dst buffer. If
+the size of src buffer is “count * dt_elem_size”, where dt_elem_size is the
+number of bytes for the data type element and count is the number of elements of
+that datatype. Assuming the result can be divided into “n” blocks, the ith block
```
For ReduceScatter we must agree:
1. Either we only support the case when count % team_size == 0 -> MPI case.
2. If we are to support other case then we need to define how the remainder (count % team_size) is distributed. Then we must document it. E.g., the remainder can be distributed evenly among first (count % team_size) ranks - ie, each of those ranks will get +1 element

manjugv replied:
Do we have use case for 2?

manjugv replied:
Discussion with Val: Let's support case 1.

vspetrov commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,150 @@
+that datatype. Assuming the result can be divided into “n” blocks, the ith block
+is placed in the receive buffer of endpoint “i”. Like other collectives, for
+this collective, the “ep” should be ordered and contiguous.
+
```
Missing description for: allgather

manjugv replied:
Done.

Sergei-Lebedev commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,169 @@
++ In the UCC\_COLL\_TYPE\_ALLTOALL collective operation, all participants
+exchange a fixed amount of the data. For a given participant, the size of data
+in src buffer is “size”, where size is dt\_elem\_size * count * num_participants.
+Here, the “count” represents the number of data elements. The "dt_elem_size"
```
i think changing to "Here, the “count” represents the number of data elements per destination" would be more clear

Sergei-Lebedev commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,169 @@
+parts (.i.e., ALLTOALL, SCATTER, GATHER, and REDUCE\_SCATTER) in that the
+location of data for the send and receive are specified by displacement arrays.
+
+**Reduction Types**: The reduction operation supported by UCC\_{ALLREDUCE,
```
add reduce scatter: "supported by UCC\_{ALLREDUCE, REDUCE, REDUCE\_SCATTER, REDUCE\_SCATTERV"

Sergei-Lebedev commented on docs/doxygen/collectives.md:
```diff
@@ -1,20 +1,169 @@
+location of data for the send and receive are specified by displacement arrays.
+
+**Reduction Types**: The reduction operation supported by UCC\_{ALLREDUCE,
+REDUCE} operation is defined by the enumeration ucc\_op\_t. The valid datatypes
```
ucc\_op\_t is no longer exists, should be ucc\_reduction\_op\_t

manjugv commented:
@Sergei-Lebedev Addressed the feedback. I hadn't touched some parts of the text where you had suggested fixes. The fixes were small enough, I fixed in this PR rather than doing another PR.
